sharding/collator.go

```diff
@@ -0,0 +1,106 @@
+package sharding
+
+import (
+	"context"
+	"fmt"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"math/big"
+)
+
+// subscribeBlockHeaders checks incoming block headers and determines if
+// we are an eligible proposer for collations. Then, it finds the pending tx's
+// from the running geth node and sorts them by descending order of gas price,
+// eliminates those that ask for too much gas, and routes them over
+// to the VMC to create a collation
+func subscribeBlockHeaders(c *Client) error {
+	headerChan := make(chan *types.Header, 16)
+
+	_, err := c.client.SubscribeNewHead(context.Background(), headerChan)
+	if err != nil {
+		return fmt.Errorf("unable to subscribe to incoming headers. %v", err)
+	}
+
+	log.Info("listening for new headers...")
+
+	for {
+		select {
+		case head := <-headerChan:
+			// Query the current state to see if we are an eligible proposer
+			log.Info(fmt.Sprintf("received new header %v", head.Number.String()))
+			// TODO: Only run this code on certain periods?
+			err := watchShards(c, head)
+			if err != nil {
+				return fmt.Errorf("unable to watch shards. %v", err)
+			}
+		}
+	}
+}
+
+// watchShards checks if we are an eligible proposer for collation for
+// the available shards in the VMC. The function calls getEligibleProposer from
+// the VMC and proposes a collation if conditions are met
+func watchShards(c *Client, head *types.Header) error {
```

Comment:
Consider changing "watch" to another verb. Perhaps checkShards since this method checks if we eligible proposer?